File: latexdraw-research/GUImutants/mutant20/net.sf.latexdraw.mutant20/src/main/net/sf/latexdraw/actions/InsertPSTCode.scala
package net.sf.latexdraw.actions

import java.text.ParseException
import org.malai.action.Action
import org.malai.undo.Undoable
import net.sf.latexdraw.badaboom.BadaboomCollector
import net.sf.latexdraw.glib.models.interfaces.IGroup
import net.sf.latexdraw.parsers.pst.parser.PSTParser
import javax.swing.JLabel
import net.sf.latexdraw.lang.LangTool
import net.sf.latexdraw.glib.models.interfaces.IShape
import net.sf.latexdraw.badaboom.BadaboomManager

/**
 * This action converts PST code into shapes and add them to the drawing.
 * <br>
 * This file is part of LaTeXDraw<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 *  LaTeXDraw is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  any later version.<br>
 * <br>
 *  LaTeXDraw is distributed without any warranty; without even the
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 *  PURPOSE. See the GNU General Public License for more details.<br>
 * <br>
 * 2013-02-19<br>
 * @author Arnaud Blouin
 * @since 3.0
 */
class InsertPSTCode extends Action with DrawingAction with Undoable {
	/** The code to parse. */
	protected var _code : Option[String] = None

	/** The status bar. */
	protected var _statusBar : Option[JLabel] = None

	/** The added shapes. */
	private var _shapes : IShape = null



	protected def doActionBody() {
		PSTParser.cleanErrors
		try {
			new PSTParser().parsePSTCode(_code.get) match {
				case Some(group) if(!group.isEmpty) =>
					if(group.size()>1)
						_shapes = group
					else _shapes = group.getShapeAt(0)

					val br = _shapes.getBottomRightPoint
					val tl = _shapes.getTopLeftPoint
					val tx = if(tl.getX<0) -tl.getX+50 else 0
					val ty = if(tl.getY<0) -tl.getY+50 else 0

					_shapes.translate(tx, ty)
					redo
					if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.36"))
				case _ => if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.33"))
			}
		}catch{
			case ex =>
				BadaboomCollector.INSTANCE.add(ex)
				if(_statusBar.isDefined) _statusBar.get.setText(LangTool.INSTANCE.getString16("LaTeXDrawFrame.34"))
			}

		PSTParser.errorLogs.foreach{str => BadaboomCollector.INSTANCE.add(new ParseException(str, -1))}
		done
	}

	override def undo() {
		if(_shapes!=null) {
			_drawing.get.removeShape(_shapes)
			_drawing.get.setModified(true)
		}
	}

	override def redo() {
		if(_shapes!=null) {
			_drawing.get.addShape(_shapes)
			_drawing.get.setModified(true)
		}
	}

	def setStatusBar(value:JLabel) { if(value!=null) _statusBar = Some(value) else _statusBar = None }

	def setCode(value:String) { if(value!=null) _code = Some(value) else _code = None }

	override def getUndoName() = "Inserting of PST code"

	override def canDo() = _code.isDefined && _drawing.isDefined

	override def hadEffect() = isDone && _shapes!=null

	override def isRegisterable() = hadEffect
}

File: src/main/java/net/sf/latexdraw/commands/InsertPSTCode.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2018 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.commands;


import java.util.Optional;
import java.util.ResourceBundle;
import javafx.scene.control.Label;
import net.sf.latexdraw.badaboom.BadaboomCollector;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.IDrawing;
import net.sf.latexdraw.models.interfaces.shape.IGroup;
import net.sf.latexdraw.models.interfaces.shape.IPoint;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.parsers.pst.PSTContext;
import net.sf.latexdraw.parsers.pst.PSTLatexdrawListener;
import net.sf.latexdraw.parsers.pst.PSTLexer;
import net.sf.latexdraw.parsers.pst.PSTParser;
import net.sf.latexdraw.util.LangService;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.RecognitionException;
import org.malai.undo.Undoable;

/**
 * This command converts PST code into shapes and add them to the drawing.
 * @author Arnaud Blouin
 */
public class InsertPSTCode extends DrawingCmdImpl implements Undoable, Modifying {
	/** The code to parse. */
	private final String code;
	/** The status bar. */
	private final Label statusBar;
	/** The added shapes. */
	private Optional<IShape> shapes;
	private final LangService lang;


	public InsertPSTCode(final String codeToInsert, final Label status, final IDrawing drawingToFill, final LangService lang) {
		super(drawingToFill);
		code = codeToInsert;
		statusBar = status;
		shapes = Optional.empty();
		this.lang = lang;

	}

	@Override
	protected void doCmdBody() {
		try {
			final PSTLatexdrawListener listener = new PSTLatexdrawListener();
			final PSTLexer lexer = new PSTLexer(CharStreams.fromString(code));
			final PSTParser parser = new PSTParser(new CommonTokenStream(lexer));
			parser.addParseListener(listener);
			parser.pstCode(new PSTContext());

			final IGroup group = ShapeFactory.INST.createGroup();
			group.getShapes().addAll(listener.flatShapes());

			if(!group.isEmpty()) {
				final IShape sh = group.size() > 1 ? group : group.getShapeAt(0);
				final IPoint tl = sh.getTopLeftPoint();
				final double tx = tl.getX() < 0d ? -tl.getX() + 50d : 0d;
				final double ty = tl.getY() < 0d ? -tl.getY() + 50d : 0d;

				shapes = Optional.of(sh);
				sh.translate(tx, ty);
				redo();

				if(statusBar != null) {
					statusBar.setText(lang.getBundle().getString("LaTeXDrawFrame.36"));
				}
			}
			parser.getInterpreter().clearDFA();
			lexer.getInterpreter().clearDFA();
		}catch(final RecognitionException ex) {
			BadaboomCollector.INSTANCE.add(ex);
			if(statusBar != null) {
				statusBar.setText(lang.getBundle().getString("LaTeXDrawFrame.34"));
			}
		}

		done();
	}

	@Override
	public void undo() {
		shapes.ifPresent(sh -> {
			drawing.removeShape(sh);
			drawing.setModified(true);
		});
	}

	@Override
	public void redo() {
		shapes.ifPresent(sh -> {
			drawing.addShape(sh);
			drawing.setModified(true);
		});
	}

	@Override
	public String getUndoName(final ResourceBundle bundle) {
		return bundle.getString("Actions.4");
	}

	@Override
	public boolean canDo() {
		return code != null && super.canDo();
	}

	@Override
	public boolean hadEffect() {
		return isDone() && shapes.isPresent();
	}

	@Override
	public RegistrationPolicy getRegistrationPolicy() {
		return hadEffect() ? RegistrationPolicy.LIMITED : RegistrationPolicy.NONE;
	}
}
